### ISSUE TITLE
公式块引用改进及锚文本的设置

### ISSUE THREAD

**Issue by Maxpiner** [siyuan-note/siyuan#660]
### 你在什么场景下需要该功能？ In what scenarios do you need this function?
理工学科相关的笔记中常常包含大量的公式及公式引用，例如“根据式(2.1.1)和式(3.2.4),可以得出”。目前思源的两种公式引用方式
(1)复制粘贴块引用，显示为id
(2)使用“((”显示为tex文本 
这两种，都需要手动修改文本，不能快捷实现例子中的效果

### 描述最优的解决方案 Describe the optimal solution
分两类场景，
1.以我个人在摘录书籍时的习惯为例，在公式中我会使用\tag{}进行公式编号以保持与书本的一致。这时希望在复制块引用或者使用((时能够自动表现为tag里的内容。例如说输入“式((6"下拉菜单自动筛选出含有\tag{6.*)的公式，确定后就自动插入编号。tag命令在mathjax和katex中都是适用的。这种方式表里统一，即tag是什么，显示的就是什么。
2.如果是以输出为目的的编写场景，则需要的应该是自动编号和ref。这方面typora采用了mathjax的自动编号（首选项/markdown 里开启该功能）+\label{}和\eqref。[typora math](https://support.typora.io/Math/)，[mathjax自动编号](http://docs.mathjax.org/en/v2.6-latest/tex.html#automatic-equation-numbering)。但是我在思源上似乎无法这么做，而且试了下发现label经常报错，重启可以消除报错但是引用仍然无法起效。这种方式表里不一定相同，label自定义、唯一且不显示，而不论是公式本身还是引用，显示的都是自动编号且无法改动。

### 描述候选解决方案 Describe the candidate solution
虽然是“候选”但是也许更重要的想法：
（1）进一步扩展，如果不局限在公式上的，思源也许可以添加给所有块自定义标识符和预置锚文本功能。
联想到latex里的label和ref可以对公式、图像、表格添加唯一的label并在其他地方引用，并显示为“图（1）”的形式。既然思源已经给块授予了唯一的id，那么也可以创造一些方式让用户在自觉有必要时手动给某个块设置label标志符（相当于id的别名，方便记忆）或者设置锚文本（这样每次引用时就直接采用这里的设置），当然两者也可以是同一个东西。锚文本可以在生成块时自动填充。在公式里可以是tag，在图表里是题注，在超级块里可以是{{{后面添加的“text”（有点类似于```后边的语言标记，虽然效果不同），再或者可以是一些更通用的，适用于所有块的方式，如右键菜单设置。同时，在输入((+关键字时也可以搜索到设置的label和锚文本。
（2）在前文中提到公式的引用以"式((”开头筛选可用公式块，但是不同人有不同习惯，比如英文下可能就是expression。进一步扩展，对于每一类块都可以设置一个独特的开头（姑且称之为“起始符”），例如“图”，“表”，“段”等等，在引用时自动筛选并按照设置的格式插入锚文本，可以自定义或者指定一个配置文件。

**Comment by 88250** [comment]
你好，我大致明白你的提议，请看一下是否和 https://github.com/siyuan-note/siyuan/issues/509 有一些共同的地方。对于公式环境的支持估计会靠后一些，因为依赖上游库 MathJax3 还有一些问题需要等待他们改进。

后面会支持[内容块命名](https://github.com/siyuan-note/siyuan/issues/595)，如果通过内容块命名引用，则默认使用该命名为锚文本，这样可能就会方便一些了，谢谢。

**Comment by Maxpiner** [comment]
感谢您的阅读和回复，下面是我的一些想法。
1. 在公式块引用方面，我的提议应该包含了 [#509](https://github.com/siyuan-note/siyuan/issues/509) 的内容。
有一点可能需要说明的是，我和509想要的其实是**公式块引用**的改进，而不是**公式引用**。公式引用是在**数学环境**中进行的，依靠label和ref，但是根据 [#40](https://github.com/siyuan-note/siyuan/issues/40) 尚未支持。而公式块引用是通过**思源自己的块引用**实现，是在引用时从tex代码中检测到关键字（`tag`或者`label`）后进行解析和填充操作，mathjax或者katex只起渲染作用。
另外，latex里label和tag起的作用是不一样的。tag仅仅用于公式中一段文本显示。而label起的是数学环境中的id作用，不会显示，可以在数学环境中被引用；显示的标号则由连续的自动编号产生。tag现在在思源中是完全没有问题的，只要针对公式块引用时生成锚文本的方式做修改就行应该能实现这部分改进功能了。
![image](https://user-images.githubusercontent.com/26576290/101322977-e1dac380-38a2-11eb-9cba-f2aaa5062d56.png)

2. 候选方案部分其实已经超出这个issue标题的内容了。其中（1）不只是内容块的命名，还有根据块的类型和内容自动识别生成命名的想法。而（2）通过`起始符+((`分类搜索块，也算是509中提到的`如果把搜索域限定在公式块中则更加快捷`的一种实现方式。

**Comment by 88250** [comment]
谢谢，我基本明白了，应该能够根据公式块内容中的 `\tag` 或者 `\label` 优先进行锚文本填充。

关于引用搜索过滤，目前已经支持按前缀进行，公式块使用 `(($`：

![image](https://user-images.githubusercontent.com/873584/101326164-d4740800-38a7-11eb-9ad4-176afb8d3733.png)

**Comment by 88250** [comment]
你好，后期会通过块属性（命名和别名）来支持搜索，这样不只公式块可以用，其他块也能用。

本提议我关闭了，感谢。

**Comment by phoenixsfly** [comment]
> 谢谢，我基本明白了，应该能够根据公式块内容中的 `\tag` 或者 `\label` 优先进行锚文本填充。
> 
> 关于引用搜索过滤，目前已经支持按前缀进行，公式块使用 `((m:`：
> 
> ![image](https://user-images.githubusercontent.com/873584/103175539-8ef29b80-48a5-11eb-9394-2d4e2708ebb3.png)

请问这个功能在最新版被移除了吗？

> 谢谢，我基本明白了，应该能够根据公式块内容中的 `\tag` 或者 `\label` 优先进行锚文本填充。
> 
> 关于引用搜索过滤，目前已经支持按前缀进行，公式块使用 `((m:`：
> 
> ![image](https://user-images.githubusercontent.com/873584/103175539-8ef29b80-48a5-11eb-9394-2d4e2708ebb3.png)

请问使用((m: 或者((h: 语法进行类型过滤搜索这一功能在最新版本中是否被移除，我在v3.1.3版本中不生效
